### ISSUE TITLE
SlevomatCodingStandard.TypeHints.ReturnTypeHint.MissingNativeTypeHint suggests void union return type for void|bool

### ISSUE THREAD

**Issue by heddn** [slevomat/coding-standard#1342]
Using recent tag 7.0.19 and following code:

```php
/**
   * @return void|bool
   */
  public function method(string $argument) {
    if ($argument !== 'foo')) {
      return FALSE;
    }
    return parent::method($entity);
  }
```

> Method \Vendor\project\Class::method() does not have native return type hint for its return value but it should be possible to add it based on @return annotation "void|bool".

However, the RFC doesn't allow combining void in union with anything else. I can't change the interface either, because the parent class is out of my control and is provided by another library project. And it is what returns void.

**Comment by kukulich** [comment]
“void|bool” does not make sense. I think the PHPCS fixer will fix it to the right “null|bool”

**Comment by heddn** [comment]
But null|bool isn't right for a void return. Where the parent doesn't return null; but rather it just returns nothing.

```php
ParentClass {

  public function method(string $argument) {
    if ($argument === 'bar')) {
      return TRUE;
    }
  }

}
```

**Comment by kukulich** [comment]
I think that parent returns “null” as well.

**Comment by heddn** [comment]
https://3v4l.org/8HABI#v8.0.17 Seems to indicate otherwise.

**Comment by ondrejmirtes** [comment]
Calling a function that returns `void` ends up with a `null` value: https://3v4l.org/um6OT

Combining `void` with something else doesn't make sense, you can't do it in native types either: https://3v4l.org/JCgFU

So your function should really have `@return null|bool`.

**Comment by heddn** [comment]
I think my concern is that the phpcs rule suggests something I can't do with PHP. If I denote `@return bool|null` then I can obviously do: `function foo(): ?bool {}`. But if `@return bool|void`, then I have to leave off the hints entirely because PHP doesn't allow `function foo(): bool|void {}`. And in the later case, the phpcs rule shouldn't suggest a native return type hint of `bool|void` is possible.

Requisite example: https://3v4l.org/4DR7o#v8.0.17

**Comment by kukulich** [comment]
The sniff does not suggest exactly `bool|void` - it only reports that it should be possible to add native type hint based on `@return bool|void`

**Comment by heddn** [comment]
But if I change it to `bool|null`, it then suggests `?bool`, which isn't possible. The only way to distinguish the void return is to use `bool|void`. And then the rule shouldn't suggest a native type hint that isn't possible, no?

**Comment by kukulich** [comment]
@heddn PHPCS works in one-file scope so it does not know anything about the parent.

**Comment by heddn** [comment]
Parent is a red herring. It only indicates that I can't change the contract. Normally I would rewrite the code to return a `?bool`. I can't because the code is upstream.

What I want is for _this_ rule to not suggest something that breaks the rules of PHP, meaning don't suggest `void` as a union with anything. It isn't possible.

**Comment by kukulich** [comment]
> What I want is for this rule to not suggest something that breaks the rules of PHP, meaning don't suggest void as a union with anything. It isn't possible.

https://github.com/slevomat/coding-standard/issues/1342#issuecomment-1070912935

**Comment by heddn** [comment]
```
 20 | ERROR | [x] Method \Vendor\project\Class::method() does not have native return
    |       |     type hint for its return value but it should be possible to add it based on
    |       |     @return annotation "bool|void".
```

**Comment by ondrejmirtes** [comment]
The correct solution is `@return bool|null` or `@return ?bool`.

**Comment by kukulich** [comment]
https://github.com/slevomat/coding-standard/issues/1342#issuecomment-1070912935
